### ISSUE TITLE
[iOS] Custom markers are always re-rendered on updates, preventing onPress from firing

### ISSUE THREAD

**Issue by IAmMarcellus** [react-native-community/react-native-maps#2860]
### Is this a bug report?

Maybe. I'm not sure if this is my error or a bug, but I've been trying to solve this specific issue for days to no avail. I think I know exactly what the problem is, but the only solution I can come up with would break other things, which is why I think there may be a bug.

### Have you read the [Installation Instructions](https://github.com/react-community/react-native-maps/blob/master/docs/installation.md)?

Yes.

### Environment

    react-native: 0.59.5
    react: 16.8.3
    react-native-maps: 0.24.2

### Steps to Reproduce

In `onRegionChangeComplete` I do `setState({region})` and in my markers' `onPress` function I also
update the state to keep track of the currently selected place and I fire an event to scroll a carousel to the same selected place.  However, if the user quickly moves the map and then selects a location (so that they press the marker after `onRegionChangeComplete` fires, but before the component finished updating), the onPress event never fires. The marker does show the callout as soon as the marker's pressed, but after a half second, the callout disappears (the markers re-render) and nothing else changes.  I think that the markers are re-rendering on every state change (i.e. every `onRegionChangeComplete`) causing the `onPress` event to not fire.

(Write your steps here:)

1. Using a map with multiple custom markers, each with their own `onPress` event, call `setState` at and update something on the component  (e.g. in `onRegionChangeComplete` or with a button)
2. Quickly press a marker before the component has updated from the last state change
3. Notice whatever was supposed to happen in the `onPress` function hasn't happened

One solution might be to use `shouldComponentUpdate,` however, the docs state that it should only be used for performance optimization and not to prevent re-renders (https://reactjs.org/docs/react-component.html#shouldcomponentupdate), but more importantly, my `componentDidUpdate` function has some conditional logic dependent on the region set in `shouldComponentUpdate,` as well as other conditional actions, so I don't want to prevent re-rendering of the entire component, just unnecessary re-rendering of the markers.

Note: I'm also using the performance optimization mentioned in https://github.com/react-native-community/react-native-maps/issues/2082 of wrapping the makers in a component that implements `shouldComponentUpdate` and `getDerivedStateFromProps`, however, I'm not entirely sure this is doing anything because it seems like the parent component is just recreating all of my optimized markers rather than using their optimizations to handle re-rendering. Also, even if I don't use a wrapped marker but a conventional custom marker, I still have the same issues.

### Expected Behavior

Setting state and then pressing a marker would always fire the `onPress` function. Additionally, it would be nice if the markers didn't render every time an ancestor component updates.

### Actual Behavior

Updating state and then quickly pressing a marker will cause the `onPress` event not to fire.  Also, the markers are re-rendered/recreated every time a parent component is updated. (I say recreated because **it seems like the markers are re-rendering without even firing `shouldComponentUpdate` or `componentDidUpdate`**).

### Reproducible Demo

(Paste the link to an example project and exact instructions to reproduce the issue.)

**Comment by jramalho** [comment]
check if the props you pass to the marker are currently updating because react-native will update PureComponent on every this.props.myProps !== nextProps.myProps or this.state.myState !== nextState.myState (objects will return true in most of the cases because the comparisson is made on the reference not the object[more info](http://adripofjavascript.com/blog/drips/object-equality-in-javascript.html)). I think you can adjust this with react lifecycle. Good luck.

**Comment by IAmMarcellus** [comment]
@jramalho That's what I thought at first, too, but that's not quite the case. It's not that the markers are updating unnecessarily, but that they're being entirely recreated. I created a wrapper component for the marker and using console logs I can see that the `componentDidMount` and `constructor` functions in the wrapper component are called every time the parent component with the map (my home screen) updates, whereas the marker wrapper's `componentDidUpdate` is never called.

You definitely pointed me in the right direction though. I'm looking through https://github.com/facebook/react/issues/8669 which describes the problem of a component with children (i.e. my map and its markers), recreating those children whenever the component's parent (my home screen) updates. I still have to look into it more, but I think it's happening because of the false equivalency that you mentioned happening with the children prop. This causes the component (my map) to re-render, which recreates the children.

**Comment by jramalho** [comment]
I suggest to you implement some lifecycles methods, the map update can lead to some performance issues. Using a Component and shouldComponentUpdate() function will prevent the updates that you dont want(maybe in the Component that is the father of the Mapview). Another suggestion us to use onRegionChangeComplete (like you dont need to update the map constantly just when you put a finger in the same, wait the action complete). 
PureComponents are cool and stuff, but they have some strange interactions with Realm-js or Maps and etc.
Maybe someone more experienced with RNMaps and RN can help, just had 4 months with React and its a love-have relationship HAHA.
btw Good luck on your implementation.

**Comment by IAmMarcellus** [comment]
I got it! Putting the map in its own PureComponent stopped the OnMarkerPress issue from happening, but the markers were still being recreated instead of re-rendering in the cases where they should have been re-rendering.  The actual problem was that I tried to use this advice: https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318 and I ended up generating new keys for the markers on ever render, causing them to be recreated every time.
Thanks for the help, @jramalho!
